rawdb/rawdb.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2024 Hemi Labs, Inc.
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package rawdb
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/juju/loggo"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+const (
+	logLevel = "INFO"
+
+	indexDir = "index"
+	dataDir  = "data"
+)
+
+var (
+	log             = loggo.GetLogger("rawdb")
+	lastFilenameKey = []byte("lastfilename")
+)
+
+func init() {
+	loggo.ConfigureLoggers(logLevel)
+}
+
+type RawDB struct {
+	mtx sync.RWMutex
+
+	home string
+
+	maxSize int64
+
+	index *leveldb.DB
+}
+
+func New(home string, maxSize int64) (*RawDB, error) {
+	log.Tracef("New")
+	defer log.Tracef("New exit")
+
+	if maxSize < 4096 {
+		return nil, fmt.Errorf("invalid max size: %v", maxSize)
+	}
+
+	return &RawDB{
+		home:    home,
+		maxSize: maxSize,
+	}, nil
+}
+
+func (r *RawDB) lastFilename() (string, error) {
+	log.Tracef("lastFilename")
+	defer log.Tracef("lastFilename exit")
+
+	return "", errors.New("not yet")
+}
+
+func (r *RawDB) Open() error {
+	log.Tracef("Open")
+	defer log.Tracef("Open exit")
+
+	err := os.MkdirAll(filepath.Join(r.home, dataDir), 0o0700)
+	if err != nil {
+		return fmt.Errorf("mkdir: %w", err)
+	}
+	r.index, err = leveldb.OpenFile(filepath.Join(r.home, indexDir), nil)
+	if err != nil {
+		return fmt.Errorf("mkdir: %w", err)
+	}
+
+	return nil
+}
+
+func (r *RawDB) Close() error {
+	log.Tracef("Close")
+	defer log.Tracef("Close exit")
+
+	err := r.index.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RawDB) Has(key []byte) (bool, error) {
+	log.Tracef("Has")
+	defer log.Tracef("Has exit")
+
+	return r.index.Has(key, nil)
+}
+
+func (r *RawDB) Insert(key, value []byte) error {
+	log.Tracef("Insert")
+	defer log.Tracef("Insert exit")
+
+	if int64(len(value)) > r.maxSize {
+		return fmt.Errorf("length exceeds maximum length: %v > %v",
+			len(value), r.maxSize)
+	}
+
+	// Assert we do not have this key stored yet.
+	if ok, err := r.index.Has(key, nil); ok {
+		return errors.New("key already exists")
+	} else if err != nil {
+		return err
+	}
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	tries := 0
+	for {
+		// This should not happen but we must ensure we aren't spinning.
+		if tries > 1 {
+			return fmt.Errorf("could not determinet last filename")
```

Comment:
nit: spelling mistake "determinet" > "determine"

Reply:
You beat me, haha. Resolved in: [`e39f694` (#201)](https://github.com/hemilabs/heminetwork/pull/201/commits/e39f694d5045e5c9a143b4d387daad31c1007619)